Sources/Networking/HTTPClient/HTTPClient.swift

```diff
@@ -582,6 +606,95 @@ private extension HTTPClient {
 
 }
 
+// MARK: - Request Retry Logic
+extension HTTPClient {
+
+    /// Evaluates whether a request should be retried and schedules a retry if necessary.
+    ///
+    /// This function checks the HTTP response status code to determine if the request should be retried.
+    /// If the retry conditions are met, it schedules the request to be retried after a backoff interval.
+    ///
+    /// - Parameters:
+    ///   - request: The original `HTTPClient.Request` that may need to be retried.
+    ///   - httpURLResponse: An optional `HTTPURLResponse` that contains the status code of the response.
+    /// - Returns: A Boolean value indicating whether the request was scheduled for a retry.
+    internal func retryRequestIfNeeded(
+        request: HTTPClient.Request,
+        httpURLResponse: HTTPURLResponse?
+    ) -> Bool {
+        guard let httpURLResponse = httpURLResponse,
+              shouldRetryRequest(withStatusCode: httpURLResponse.httpStatusCode) else { return false }
+
+        // At this point, retryCount hasn't been incremented yet, so we'll need to do it early here
+        // to determine if another retry is appropriate.
+        let nextRetryCount = request.retryCount + 1
+
+        guard nextRetryCount <= self.retryBackoffIntervals.count else {
+            Logger.error(
+                NetworkStrings.api_request_failed_all_retries(
+                    httpMethod: request.method.httpMethod,
+                    path: request.path,
+                    retryCount: request.retryCount
+                )
+            )
+            return false
+        }
+
+        let retryBackoffInterval: TimeInterval = calculateRetryBackoffTime(
+            forResponse: httpURLResponse,
+            retryCount: nextRetryCount
+        )
+
+        Logger.debug(
+            NetworkStrings.api_request_queued_for_retry(
+                httpMethod: request.method.httpMethod,
+                path: request.path,
+                backoffInterval: retryBackoffInterval
+            )
+        )
+        self.operationDispatcher.dispatchOnWorkerThread(delay: .timeInterval(retryBackoffInterval)) {
+            let retriedRequest = request.retriedRequest()
+            self.state.modify {
+                $0.queuedRequests.insert(retriedRequest, at: 0)
+            }
+        }
+        return true
+    }
+
+    internal func shouldRetryRequest(withStatusCode statusCode: HTTPStatusCode) -> Bool {
+        return self.retriableStatusCodes.contains(statusCode)
+    }
+
+    internal func calculateRetryBackoffTime(
+        forResponse httpURLResponse: HTTPURLResponse,
+        retryCount: UInt
+    ) -> TimeInterval {
+        // Use the retry after value from the backend if present
+        if let retryAfterHeaderValue = httpURLResponse.allHeaderFields[ResponseHeader.retryAfter.rawValue] as? String {
+            if let retryAfterMS = Double(retryAfterHeaderValue) {
```

Comment:
could this be in the same `if`?
```
if let retryAfterHeaderValue = httpURLResponse.allHeaderFields[ResponseHeader.retryAfter.rawValue] as? String, let retryAfterMS = Double(retryAfterHeaderValue) {
```

Reply:
Nice catch! Always nice to have fewer nested conditions 😄